### PR TITLE
[Fix] Optimization in the consistency check to skip recent deleted flows 

### LINKED ISSUE #69: Consistency check possible removal of legitime flow
I was running a few tests in disruptive scenarios (i.e., scenarios with link failures/flaps) and with a bunch of EVCs (around 100 EVCs) and everything worked as expected, except for one test in which one EVC didn't work as expected (after the failed link return to its normal state - UP - the end user started seen packet loss).

The test consists of the following workflow:

1. 100 EVCs are created using dynamic path
2. After the creation of the EVCs, we start a PING connectivity test between the endpoints of each EVC
3. A few seconds later, the link `00:00:00:00:00:00:00:01:11--00:00:00:00:00:00:00:06:11` goes down
4. 60 seconds later, the link `00:00:00:00:00:00:00:01:11--00:00:00:00:00:00:00:06:11` come up
5. ping keeps running for more 30 seconds and the test finishes.

Due to a problem in the way link down events are being handled by mef_eline, the EVC does not actually gets moved to the backup path during the link failure. However, after the link has come UP again, mef_eline runs a reprovisioning on the EVC and setup new flows (actually, at this point mef_eline was still running the loop over the EVCs affected by the link failure.. so that, the reprovisioning of the EVC will be through the same path, but now using different service vlans).

Below is a sequence of filtered log messages showing the behavior in terms of mef_eline (evc creation and reprovisioning), of_core (interface down and up) and flow_manager (flows creation, removal and consistency check):

```
# zcat /var/log/kytos/kytos.log-20220203012301.gz | egrep "12319557280506940481|f7dc1326b18c41|Consistency check:|Modified"
##
## Note: below is the EVC creation
##
2022-02-03 06:27:16,977 - INFO [kytos.napps.kytos/flow_manager] (Thread-916) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319557280506940481, 'cookie_mask': 18446744073709551615}], 'force': True}
2022-02-03 06:27:17,202 - INFO [kytos.napps.kytos/flow_manager] (Thread-917) Send FlowMod from request dpid: 00:00:00:00:00:00:00:06, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319557280506940481, 'cookie_mask': 18446744073709551615}], 'force': True}
2022-02-03 06:27:17,259 - INFO [kytos.napps.kytos/mef_eline] (Thread-915) EVC f7dc1326b18c41 was synced to the storehouse.
2022-02-03 06:27:17,418 - INFO [kytos.napps.kytos/flow_manager] (Thread-919) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 16, 'dl_vlan': 1149}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 859}, {'action_type': 'output', 'port': 11}]}, {'match': {'in_port': 11, 'dl_vlan': 859}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 1149}, {'action_type': 'output', 'port': 16}]}], 'force': False}
2022-02-03 06:27:17,870 - INFO [kytos.napps.kytos/flow_manager] (Thread-920) Send FlowMod from request dpid: 00:00:00:00:00:00:00:06, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 21, 'dl_vlan': 1149}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 859}, {'action_type': 'output', 'port': 11}]}, {'match': {'in_port': 11, 'dl_vlan': 859}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 1149}, {'action_type': 'output', 'port': 21}]}], 'force': False}
2022-02-03 06:27:18,451 - INFO [kytos.napps.kytos/mef_eline] (Thread-915) EVC f7dc1326b18c41 was synced to the storehouse.
2022-02-03 06:27:18,466 - INFO [kytos.napps.kytos/mef_eline] (Thread-915) EVC(f7dc1326b18c41, my evc 1149) was deployed.
##
## Note: Link goes down
##
2022-02-03 06:28:51,760 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_31) Modified Interface('novi_port_11', 11, Switch('00:00:00:00:00:00:00:01')) 00:00:00:00:00:00:00:01:11
2022-02-03 06:28:51,821 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_1) Modified Interface('novi_port_11', 11, Switch('00:00:00:00:00:00:00:06')) 00:00:00:00:00:00:00:06:11
...
##
## Note: Link come up again
##
2022-02-03 06:29:55,311 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_8) Modified Interface('novi_port_11', 11, Switch('00:00:00:00:00:00:00:06')) 00:00:00:00:00:00:00:06:11
2022-02-03 06:29:55,416 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_21) Modified Interface('novi_port_11', 11, Switch('00:00:00:00:00:00:00:01')) 00:00:00:00:00:00:00:01:11
##
## Note: mef_eline path  reprovision for the EVC
##
2022-02-03 06:30:38,073 - INFO [kytos.napps.kytos/flow_manager] (Thread-1265) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319557280506940481, 'cookie_mask': 18446744073709551615}], 'force': True}
2022-02-03 06:30:38,298 - INFO [kytos.napps.kytos/flow_manager] (Thread-1266) Send FlowMod from request dpid: 00:00:00:00:00:00:00:06, command: delete, force: True, flows_dict: {'flows': [{'cookie': 12319557280506940481, 'cookie_mask': 18446744073709551615}], 'force': True}
2022-02-03 06:30:38,591 - INFO [kytos.napps.kytos/mef_eline] (ThreadPoolExecutor-0_13) EVC f7dc1326b18c41 was synced to the storehouse.
2022-02-03 06:30:38,779 - INFO [kytos.napps.kytos/flow_manager] (Thread-1268) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 16, 'dl_vlan': 1149}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 1455}, {'action_type': 'output', 'port': 11}]}, {'match': {'in_port': 11, 'dl_vlan': 1455}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 1149}, {'action_type': 'output', 'port': 16}]}], 'force': False}
2022-02-03 06:30:39,071 - INFO [kytos.napps.kytos/flow_manager] (Thread-1269) Send FlowMod from request dpid: 00:00:00:00:00:00:00:06, command: add, force: False, flows_dict: {'flows': [{'match': {'in_port': 21, 'dl_vlan': 1149}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'action_type': 'set_vlan', 'vlan_id': 1455}, {'action_type': 'output', 'port': 11}]}, {'match': {'in_port': 11, 'dl_vlan': 1455}, 'cookie': 12319557280506940481, 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'action_type': 'set_vlan', 'vlan_id': 1149}, {'action_type': 'output', 'port': 21}]}], 'force': False}
2022-02-03 06:30:39,964 - INFO [kytos.napps.kytos/mef_eline] (ThreadPoolExecutor-0_13) EVC f7dc1326b18c41 was synced to the storehouse.
2022-02-03 06:30:40,122 - INFO [kytos.napps.kytos/mef_eline] (ThreadPoolExecutor-0_13) EVC(f7dc1326b18c41, my evc 1149) was deployed.
##
## Note: flow_manager consistency routine
##
2022-02-03 06:30:41,051 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: alien flow on switch 00:00:00:00:00:00:00:06
2022-02-03 06:30:41,108 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Flow forwarded to switch 00:00:00:00:00:00:00:06 to be deleted. Flow: {'flows': [{'switch': '00:00:00:00:00:00:00:06', 'table_id': 0, 'match': {'in_port': 21, 'dl_vlan': 1149}, 'priority': 32768, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12319557280506940481, 'id': '13ef7612927e97f13d3f1942ff3cdf71', 'stats': {'byte_count': 320166, 'duration_sec': 198, 'duration_nsec': 426511681, 'packet_count': 2541}, 'cookie_mask': 0, 'instructions': [{'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 's'}, {'vlan_id': 859, 'action_type': 'set_vlan'}, {'port': 11, 'action_type': 'output'}]}]}]}
2022-02-03 06:30:41,146 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: alien flow on switch 00:00:00:00:00:00:00:06
2022-02-03 06:30:41,305 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Flow forwarded to switch 00:00:00:00:00:00:00:06 to be deleted. Flow: {'flows': [{'switch': '00:00:00:00:00:00:00:06', 'table_id': 0, 'match': {'in_port': 11, 'dl_vlan': 859}, 'priority': 32768, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12319557280506940481, 'id': '9dc8c461eb1e6befea3cc50e303f8d58', 'stats': {'byte_count': 320166, 'duration_sec': 198, 'duration_nsec': 369709443, 'packet_count': 2541}, 'cookie_mask': 0, 'instructions': [{'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'action_type': 'push_vlan', 'tag_type': 'c'}, {'vlan_id': 1149, 'action_type': 'set_vlan'}, {'port': 21, 'action_type': 'output'}]}]}]}
2022-02-03 06:30:41,349 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: alien flow on switch 00:00:00:00:00:00:00:06
2022-02-03 06:30:41,460 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: alien flow on switch 00:00:00:00:00:00:00:06
2022-02-03 06:30:43,314 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_32) Consistency check: missing flow on switch 00:00:00:00:00:00:00:06.
```

The above log messages show the situation described before. It is important to notice the last set of messages: the ones related to the "flow_manager consistency routine". As discussed before, mef_eline takes longer to react to the link failure for this EVC, and it ends up using the same (recovered) path but with different service-vlans. The customer VLAN is 1149, and previous service VLAN was 859 while the new service VLAN is  1455. Everything should be working at the time `2022-02-03 06:30:40,122` (when the evc was reprovisioned). However, at `2022-02-03 06:30:41,051 ` the flow_manager consistency routine identifies alien flows at the switch 00:00:00:00:00:00:00:06 and sends out a FlowMod Remove. The problem is at the first flow removal, where the match conditions are the same as the correct flow at the switch. Unfortunately, the there is no evidence in the logs about the consistency check fixing this mistake. However, a dump of the switch's flow_table at the end of the test reveals that indeed switch 00:06 have one flow missing:
```
$ grep -A 10 f7dc1326b18c41 dump-flowtable-20220203012301-xxx.xxx.xxx.229
        Cookie           = aaf7dc1326b18c41
        Send_flow_rem    = true
        Persistent       = false
        [MATCHFIELDS]
            OFPXMT_OFB_IN_PORT = 11
            OFPXMT_OFB_VLAN_VID = 1455(OFPVID_PRESENT=1)
        [INSTRUCTIONS]
            [OFPIT_APPLY_ACTIONS]
                 [ACTIONS]
                    [OFPAT_POP_VLAN]
                    [OFPAT_PUSH_VLAN]
```

I've tried to extract the flow.persistency objects from storehouse for this execution, however due to another issue (#67) it was not readable.

### PR DESCRIPTION
**Pull request by viniarck**
Fixes #69 

### Description of the change

This PR introduced an optimization opportunity that @italovalcy [has found on this comment](https://github.com/kytos-ng/flow_manager/issues/69#issuecomment-1030370568), so, now in the same way that consistency tries to skip a pass if a recent flow was added it's doing the same thing for recent deleted flows. I've also cherry picked some commits from PR #45, that I worked sometime ago for archived flows, however, the main difference to avoid putting more pressure on store house is that I only kept the parts we'd need for this feature with storage in memory, so no additional IO/DB pressure is expected. I think it's a win-win situation since now we've got the archived flows partly landed as well, and then when we have a new NoSQL backend it's a matter of writing and reading from there. 

### Release notes 

`[2022.1.2] - 2022-02-07`

#### Changed

- Adapted consistency check to skip recent deleted flows
- Extracted ``is_recent_flow`` static method
- Changed ``_del_matched_flows_store`` to also archive flows
- Changed consistency check to also archive alien flows

#### Added

- Added archived_flows and its lock to store in memory
- Added ``_add_Archived_flows`` method

#### Local tests

I've also tried to reproduce the issue with the commands and scripts that @italovalcy  provided on [this comment](https://github.com/kytos-ng/flow_manager/issues/69#issuecomment-1030370568), in three executions, no alien deleted consistency checks were observed. Also, exploring `archived_flows` in memory, the 200 flow entries were there, all of them with `delete` reason:

```
kytos $> len(controller.napps[('kytos', 'flow_manager')].archived_flows['00:00:00:00:00:00:00:01'])
Out[8]: 200


kytos $> set([v['reason'] for v in controller.napps[('kytos', 'flow_manager')].archived_flows['00:00:00:00:00:00:00:01'].values()])
Out[9]: {'delete'}
```

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -21,6 +21,23 @@ Removed
 Security
 ========
 
+[2022.1.2] - 2022-02-07
```
Comment (viniarck) [inline CHANGELOG.rst:24]: Intentionally skipped version `2022.1.1`, assuming that it'll land on this PR #68

File: setup.cfg [hunk 2]
```diff
@@ -6,7 +6,7 @@ add-ignore = D105
 
 [yala]
 radon mi args = --min C
-pylint args = --disable==too-many-arguments,too-many-locals,too-few-public-methods,too-many-instance-attributes,no-else-return,dangerous-default-value,duplicate-code,raise-missing-from,too-many-arguments,too-many-public-methods --ignored-modules=napps.kytos.topology
+pylint args = --disable==too-many-arguments,too-many-locals,too-few-public-methods,too-many-instance-attributes,no-else-return,dangerous-default-value,duplicate-code,raise-missing-from,too-many-arguments,too-many-public-methods,too-many-lines --ignored-modules=napps.kytos.topology
```
Comment (viniarck) [inline setup.cfg:9]: I had to temporarily add `too-many-lines` to be ignored for now, however, we've got on our radar to do some refactoring to split these methods in new modules, for instance this issue here as well https://github.com/kytos-ng/flow_manager/issues/11

**Comment by viniarck** [comment]
For the record, here's the result of end-to-end tests with this branch:

```
+ python3 -m pytest tests/
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /builds/amlight/kytos-end-to-end-tester/kytos-end-to-end-tests
plugins: timeout-2.0.2
collected 189 items
tests/test_e2e_01_kytos_startup.py ..                                    [  1%]
tests/test_e2e_05_topology.py .................                          [ 10%]
tests/test_e2e_10_mef_eline.py .........X.......x....x.....              [ 24%]
tests/test_e2e_11_mef_eline.py .....                                     [ 27%]
tests/test_e2e_12_mef_eline.py .....xx.                                  [ 31%]
tests/test_e2e_13_mef_eline.py .....xxxx......xxxx.XXxX.xxxx..x......... [ 53%]
...                                                                      [ 55%]
tests/test_e2e_14_mef_eline.py x                                         [ 55%]
tests/test_e2e_15_maintenance.py ........................                [ 68%]
tests/test_e2e_20_flow_manager.py ................                        [76%]
tests/test_e2e_21_flow_manager.py ..                                      [77%]
tests/test_e2e_22_flow_manager.py ...............                        [85%]
tests/test_e2e_23_flow_manager.py X......................                [ 97%]
tests/test_e2e_30_of_lldp.py ....                                        [100%]
=============================== warnings summary ===============================
------------------------------- start/stop times -------------------------------
==== 165 passed, 19 xfailed, 5 xpassed, 560 warnings in 9798.94s (2:43:18) =====

**Review by italovalcy — APPROVED** [review]
Looks good to me. It is very nice that you were able to quickly design a solution for this optimization. The set of tests presented in the PR description shows the solution is secure and effective. Thanks for fixing this, @viniarck

**Comment by viniarck** [comment]
> Looks good to me. It is very nice that you were able to quickly design a solution for this optimization. The set of tests presented in the PR description shows the solution is secure and effective. Thanks for fixing this, @viniarck

Awesome. Thanks for reviewing, Italo.

**Pull request merged**